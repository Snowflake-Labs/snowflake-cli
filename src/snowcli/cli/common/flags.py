from __future__ import annotations

import typer
from snowcli.utils import check_for_connection

DEFAULT_CONTEXT_SETTINGS = {"help_option_names": ["-h", "--help"]}


ConnectionOption = typer.Option(
    "dev",
    "-c",
    "--connection",
    "--environment",
    help="Connection / environment name",
    callback=check_for_connection,
    is_eager=True,
)

AccountOption = typer.Option(
    None,
    "-a",
    "--accountname",
    "--account",
    prompt="Snowflake account",
    help="Name assigned to your Snowflake account.",
)

UserOption = typer.Option(
    None,
    "-u",
    "--username",
    "--user",
    prompt="Snowflake username",
    help="Username to connect to Snowflake.",
)

PasswordOption = typer.Option(
    None,
    "-p",
    "--password",
    prompt="Snowflake password",
    help="Snowflake password",
    hide_input=True,
)

DatabaseOption = typer.Option(
    None,
    "-d",
    "--dbname",
    "--database",
    prompt="Database name",
    help="Database to use.",
)

SchemaOption = typer.Option(
    None,
    "-s",
    "--schemaname",
    "--schema",
    prompt="Database schema",
    help=" Schema in the database to use.",
)

RoleOption = typer.Option(
    None, "-r", "--rolename", "--role", prompt="Role name", help="Role to be used"
)

WarehouseOption = typer.Option(
    None, "-w", "--warehouse", prompt="Warehouse name", help="Warehouse to use."
)
