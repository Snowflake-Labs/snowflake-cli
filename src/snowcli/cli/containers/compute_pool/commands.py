import typer
from snowcli.api.output.decorators import with_output
from snowcli.api.output.types import CommandResult, SingleQueryResult
from snowcli.cli.common.decorators import global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.containers.compute_pool.manager import ComputePoolManager

app = typer.Typer(
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    name="pool",
    help="Manages compute pools.",
)


@app.command()
@with_output
@global_options_with_connection
def create(
    name: str = typer.Option(..., "--name", help="Name of the compute pool."),
    num_instances: int = typer.Option(
        ..., "--num", help="Number of compute pool instances."
    ),
    instance_family: str = typer.Option(
        ...,
        "--family",
        help="Name of the instance family. For more information about instance families, refer to the SQL CREATE COMPUTE POOL command.",
    ),
    **options,
) -> CommandResult:
    """
    Creates a compute pool with a specified number of instances.
    """
    cursor = ComputePoolManager().create(
        pool_name=name, num_instances=num_instances, instance_family=instance_family
    )
    return SingleQueryResult(cursor)


@app.command()
@with_output
@global_options_with_connection
def stop(
    name: str = typer.Argument(..., help="Name of the compute pool."), **options
) -> CommandResult:
    """
    Stops a compute pool and deletes all services running on the pool.
    """
    cursor = ComputePoolManager().stop(pool_name=name)
    return SingleQueryResult(cursor)
