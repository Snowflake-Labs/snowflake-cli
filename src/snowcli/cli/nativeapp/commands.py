from typing import Optional

import logging
import typer

from snowcli.cli.common.decorators import global_options, global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.output.decorators import with_output

from snowcli.cli.nativeapp.init import nativeapp_init
from snowcli.cli.nativeapp.manager import NativeAppManager

from snowcli.output.types import CommandResult, MessageResult

app = typer.Typer(
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    hidden=True,
    name="app",
    help="Manage Native Apps in Snowflake",
)

log = logging.getLogger(__name__)

ProjectArgument = typer.Option(
    None,
    "-p",
    "--project",
    help="Path where the Native Apps project resides. Defaults to current working directory",
    show_default=False,
)


@app.command("init")
@with_output
@global_options
def app_init(
    name: str = typer.Argument(
        ..., help="Name of the Native Apps project to be initiated."
    ),
    git_url: str = typer.Option(
        None,
        help="A git URL to use as template for the Native Apps project. Example: https://github.com/Snowflake-Labs/native-apps-templates.git for all official Snowflake templates.",
    ),
    template: str = typer.Option(
        None,
        help="A specific directory within the git URL to use as template for the Native Apps project. Example: Default is native-app-basic if --git-url is https://github.com/Snowflake-Labs/native-apps-templates.git, and None if any other --git-url.",
    ),
    **options,
) -> CommandResult:
    """
    Initializes a Native Apps project, optionally with a --git-url and a --template.
    """
    nativeapp_init(name, git_url, template)
    return MessageResult(
        f"Native Apps project {name} has been created in your local directory."
    )


@app.command("bundle", hidden=True)
@with_output
@global_options_with_connection
def app_bundle(
    project_path: Optional[str] = ProjectArgument,
    **options,
) -> CommandResult:
    """
    Prepares a local folder with configured app artifacts.
    """
    manager = NativeAppManager(project_path)
    manager.build_bundle()
    return MessageResult(f"Bundle generated at {manager.deploy_root}")


@app.command("run")
@with_output
@global_options
def app_run(
    project_path: Optional[str] = ProjectArgument,
    **options,
) -> CommandResult:
    """
    Creates an application package in your Snowflake account and uploads code files to its stage.
    """
    manager = NativeAppManager(project_path)
    manager.build_bundle()
    manager.app_run()
    return MessageResult(
        f"Application Package is now active in your Snowflake account!"
    )


@app.command("teardown")
@with_output
@global_options_with_connection
def app_teardown(
    project_path: Optional[str] = ProjectArgument,
    **options,
) -> CommandResult:
    """
    Drops an application and an application package as defined in the project definition file.
    """
    manager = NativeAppManager(project_path)
    manager.teardown()
    return MessageResult(f"Teardown is now complete.")
