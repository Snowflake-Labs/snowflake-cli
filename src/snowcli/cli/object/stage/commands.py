from __future__ import annotations

from pathlib import Path

import typer
from snowcli.cli.common.decorators import global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.object.stage.diff import DiffResult
from snowcli.cli.object.stage.manager import StageManager
from snowcli.output.decorators import with_output
from snowcli.output.types import (
    CommandResult,
    ObjectResult,
    QueryResult,
    SingleQueryResult,
)

app = typer.Typer(
    name="stage",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Manages stages.",
)

StageNameArgument = typer.Argument(..., help="Name of the stage.")


@app.command("list")
@with_output
@global_options_with_connection
def stage_list(stage_name: str = StageNameArgument, **options) -> CommandResult:
    """
    Lists the stage contents.
    """
    cursor = StageManager().list(stage_name=stage_name)
    return QueryResult(cursor)


@app.command("get")
@with_output
@global_options_with_connection
def stage_get(
    stage_name: str = StageNameArgument,
    path: Path = typer.Argument(
        Path.cwd(),
        exists=False,
        file_okay=True,
        dir_okay=True,
        writable=True,
        resolve_path=True,
        help="Directory location to store downloaded files. If omitted, the uploads files in the active directory.",
    ),
    **options,
) -> CommandResult:
    """
    Downloads all files from a stage to a local directory.
    """
    cursor = StageManager().get(stage_name=stage_name, dest_path=path)
    return SingleQueryResult(cursor)


@app.command("put")
@with_output
@global_options_with_connection
def stage_put(
    path: Path = typer.Argument(
        ...,
        exists=False,
        file_okay=True,
        dir_okay=True,
        writable=True,
        resolve_path=True,
        help="File or directory to upload to stage. You can use the `*` wildcard in the path, like `folder/*.csv`. If a path contains `*.`, you must enclose the path in quotes.",
    ),
    stage_name: str = StageNameArgument,
    overwrite: bool = typer.Option(
        False,
        help="Overwrites existing files in the stage.",
    ),
    parallel: int = typer.Option(
        4,
        help="Number of parallel threads to use when uploading files. Default: 4.",
    ),
    **options,
) -> CommandResult:
    """
    Uploads files to a stage from a local client.
    """
    manager = StageManager()
    local_path = str(path) + "/*" if path.is_dir() else str(path)

    cursor = manager.put(
        local_path=local_path,
        stage_path=stage_name,
        overwrite=overwrite,
        parallel=parallel,
    )
    return SingleQueryResult(cursor)


@app.command("create")
@with_output
@global_options_with_connection
def stage_create(stage_name: str = StageNameArgument, **options) -> CommandResult:
    """
    Creates a named stage if it does not already exist.
    """
    cursor = StageManager().create(stage_name=stage_name)
    return SingleQueryResult(cursor)


@app.command("remove")
@with_output
@global_options_with_connection
def stage_remove(
    stage_name: str = StageNameArgument,
    file_name: str = typer.Argument(..., help="Name of the file to remove."),
    **options,
) -> CommandResult:
    """
    Removes a file from a stage.
    """

    cursor = StageManager().remove(stage_name=stage_name, path=file_name)
    return SingleQueryResult(cursor)


@app.command("diff", hidden=True)
@with_output
@global_options_with_connection
def stage_diff(
    stage_name: str = typer.Argument(None, help="Fully qualified name of a stage"),
    folder_name: str = typer.Argument(None, help="Path to local folder"),
    **options,
) -> ObjectResult:
    """
    Diffs a stage with a local folder.
    """
    diff: DiffResult = stage_diff(Path(folder_name), stage_name)
    return ObjectResult(
        {
            "only on local": diff.only_local,
            "only on stage": diff.only_on_stage,
            "modified/unknown": diff.different,
            "identical": diff.identical,
        }
    )
