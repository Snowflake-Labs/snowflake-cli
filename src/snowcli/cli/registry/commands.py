import json
import requests
import sys
import typer
from typing import Optional

from snowcli.cli.common.decorators import global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS, ConnectionOption
from snowcli.cli.registry.manager import RegistryManager

app = typer.Typer(
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    name="registry",
    help="Manages Snowpark registries.",
)


@app.command("token")
@global_options_with_connection
def token(**options):
    """
    Gets the token from environment to use for authenticating with the registry.
    """
    sys.stdout.write(json.dumps(RegistryManager().get_token()))


@app.command("list-images")
@global_options_with_connection
def list_images(
    repo_name: str = typer.Option(
        ...,
        "--repository_name",
        "-r",
        help="Name of the image repository as seen in `show image repositories`",
    ),
    **options,
):
    database = RegistryManager().get_database()
    schema = RegistryManager().get_schema()
    url = RegistryManager().get_repository_url(repo_name)
    api_url = RegistryManager().get_repository_api_url(url)
    bearer_login = RegistryManager().login_to_registry(api_url)

    repos = []
    query: Optional[str] = f"{api_url}/_catalog?n=10"

    while query is not None:
        # Make paginated catalog requests
        response = requests.get(
            query, headers={"Authorization": f"Bearer {bearer_login}"}
        )

        if response.status_code != 200:
            print("Call to the registry failed", response.text)

        data = json.loads(response.text)
        if "repositories" in data:
            repos.extend(data["repositories"])

        if "Link" in response.headers:
            # There are more results
            query = f"{api_url}/_catalog?n=10&last={repos[-1]}"
        else:
            query = None

    sys.stdout.write("Images in this repository:\n\n")
    for repo in repos:
        prefix = f"{database}/{schema}/{repo_name}/"
        repo = repo.replace("baserepo/", prefix)

        sys.stdout.write(f"{repo}\n")


@app.command("list-tags")
@global_options_with_connection
def list_tags(
    repo_name: str = typer.Option(
        ...,
        "--repository_name",
        "-r",
        help="Name of the image repository as seen in `show image repositories`",
    ),
    image_name: str = typer.Option(
        ...,
        "--image_name",
        "-i",
        help="Name of the image as shown in the output of list-images",
    ),
    **options,
):

    url = RegistryManager().get_repository_url(repo_name)
    api_url = RegistryManager().get_repository_api_url(url)
    bearer_login = RegistryManager().login_to_registry(api_url)

    repo_name = image_name.split("/")[2]
    image_realname = "/".join(image_name.split("/")[3:])

    tags = []
    query: Optional[str] = f"{api_url}/{image_realname}/tags/list?n=10"

    while query is not None:
        # Make paginated catalog requests
        response = requests.get(
            query, headers={"Authorization": f"Bearer {bearer_login}"}
        )

        if response.status_code != 200:
            print("Call to the registry failed", response.text)

        data = json.loads(response.text)
        if "tags" in data:
            tags.extend(data["tags"])

        if "Link" in response.headers:
            # There are more results
            query = f"{api_url}/{image_realname}/tags/list?n=10&last={tags[-1]}"
        else:
            query = None

    sys.stdout.write("Tags for this image:\n\n")
    for tag in tags:
        sys.stdout.write(f"{image_name}:{tag}\n")
