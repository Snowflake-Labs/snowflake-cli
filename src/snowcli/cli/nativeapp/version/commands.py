import logging
from typing import Optional

import typer
from click import MissingParameter
from snowcli.cli.common.cli_global_context import cli_context
from snowcli.cli.common.decorators import (
    global_options_with_connection,
    with_project_definition,
)
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.nativeapp.policy import AllowAlwaysPolicy, AlwaysAskPolicy
from snowcli.cli.nativeapp.version.version_processor import (
    NativeAppVersionCreateProcessor,
    NativeAppVersionDropProcessor,
)
from snowcli.output.decorators import with_output
from snowcli.output.types import CommandResult, MessageResult

app = typer.Typer(
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    name="version",
    help="Manage Native App Pkg versions in Snowflake",
)

log = logging.getLogger(__name__)


@app.command()
@with_output
@global_options_with_connection
def create(
    version: Optional[str] = typer.Argument(
        None,
        help=f"""The identifier or 'version string' of the version you would like to create a version and/or patch for.
        Defaults to undefined, which means the CLI will use the version, if present, in the manifest.yml.""",
    ),
    patch: Optional[str] = typer.Option(
        None,
        "--patch",
        "-p",
        help=f"""The patch number you would like to create for an existing version.
        Defaults to undefined if it is not set, which means the CLI will either use the version, if present, in the manifest.yml,
        or auto-generate the patch number.""",
    ),
    force: Optional[bool] = typer.Option(
        False,
        "--force",
        help="Defaults to False. Passing in --force turns this to True, i.e. we will implicitly respond “yes” to any prompts that come up.",
        is_flag=True,
    ),
    **options,
) -> CommandResult:
    """
    Adds a new patch to the provided version for your application package. If the version does not exist, creates a version with patch 0.
    """
    if version is None and patch is not None:
        raise MissingParameter("Cannot provide a patch without version!")

    if force:
        policy = AllowAlwaysPolicy()
    else:
        policy = AlwaysAskPolicy()

    processor = NativeAppVersionCreateProcessor(
        project_definition=cli_context.project_definition,
        project_root=cli_context.project_root,
    )
    processor.process(version, patch, policy)
    return MessageResult(f"Version create is now complete.")


@app.command()
@with_output
@with_project_definition("native_app")
@global_options_with_connection
def drop(
    version: Optional[str] = typer.Argument(
        None,
        help="Version of the app package that you would like to drop. Defaults to the version specified in the manifest.yml.",
    ),
    force: Optional[bool] = typer.Option(
        False,
        "--force",
        help="Defaults to False. Passing in --force turns this to True, i.e. we will implicitly respond “yes” to any prompts that come up.",
        is_flag=True,
    ),
    **options,
) -> CommandResult:
    """
    Drops a version associated with your application package. Version can either be passed in as an argument to the command or read from the manifest.yml file.
    Dropping patches is not allowed.
    """
    if force:
        policy = AllowAlwaysPolicy()
    else:
        policy = AlwaysAskPolicy()

    processor = NativeAppVersionDropProcessor(
        project_definition=cli_context.project_definition,
        project_root=cli_context.project_root,
    )
    processor.process(version, policy)
    return MessageResult(f"Version drop is now complete.")
