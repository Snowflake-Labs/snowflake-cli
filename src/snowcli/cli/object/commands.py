from __future__ import annotations

import typer

from snowcli.cli.common.decorators import global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.object.show.manager import ObjectManager
from snowcli.cli.object.utils import ObjectType
from snowcli.output.decorators import with_output
from snowcli.output.types import QueryResult

from snowcli.cli.object.stage.commands import app as stage_app

app = typer.Typer(
    name="object",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Manages Snowflake objects like warehouses and stages",
)
app.add_typer(stage_app)  # type: ignore

ObjectArgument = typer.Argument(None, help="Type of object")
LikeOption = typer.Option(
    "%%",
    "--like",
    "-l",
    help='Regular expression for filtering the functions by name. For example, `list --like "my%"` lists all functions in the **dev** (default) environment that begin with “my”.',
)


@app.command()
@with_output
@global_options_with_connection
def show(
    object_type: ObjectType = ObjectArgument,
    like: str = LikeOption,
    **options,
):
    "Lists all avaiable Snowflake objects of given type"
    return QueryResult(ObjectManager().show(object_type, like))

@app.command()
@with_output
@global_options_with_connection
def drop(object_type: ObjectType = ObjectArgument, object_name: str = ""):
    "Drops Snowflake object of given name and type"
    return QueryResult(ObjectManager().drop(object_type,object_name))
