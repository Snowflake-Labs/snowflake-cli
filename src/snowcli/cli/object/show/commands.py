from __future__ import annotations

import typer

from snowcli.cli.common.decorators import global_options_with_connection
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.object.show.manager import ObjectManager
from snowcli.cli.object.utils import ObjectType
from snowcli.output.decorators import with_output
from snowcli.output.types import QueryResult

app = typer.Typer(
    name="show",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Lists existing objects of given type",
)

LikeOption = typer.Option(
    "%%",
    "--like",
    "-l",
    help='Regular expression for filtering the functions by name. For example, `list --like "my%"` lists all functions in the **dev** (default) environment that begin with “my”.',
)


@app.command("warehouses")
@with_output
@global_options_with_connection
def warehouse_status(like: str, **options):
    """
    Shows existing warehouses
    """
    cursor = ObjectManager().show(ObjectType.WAREHOUSE, like)
    return QueryResult(cursor)
