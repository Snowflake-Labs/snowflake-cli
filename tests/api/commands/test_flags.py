import pytest
from snowflake.cli.api.commands.flags import (
    PLAIN_PASSWORD_MSG,
    OverrideableOption,
    PasswordOption,
)
from typer import Typer
from typer.testing import CliRunner


def test_format(runner, snapshot):
    result = runner.invoke(
        ["object", "stage", "list", "stage_name", "--format", "invalid_format"]
    )

    assert result.output == snapshot


def test_password_flag():
    app = Typer()

    @app.command()
    def _(password: str = PasswordOption):
        return "ok"

    runner = CliRunner()
    result = runner.invoke(app, ["--password", "dummy"], catch_exceptions=False)
    assert result.exit_code == 0
    assert PLAIN_PASSWORD_MSG in result.output


def test_overrideable_option_is_overrideable(snapshot):
    original = OverrideableOption(1, "--option", help="original help")
    app = Typer()

    @app.command()
    def _(option: int = OverrideableOption(default=2, help="new help")):
        return "ok"

    runner = CliRunner()
    result = runner.invoke(app, ["--help"], catch_exceptions=False)
    assert result.exit_code == 0
    assert result.output == snapshot


@pytest.mark.parametrize("set1, set2", [(False, False), (False, True), (True, False)])
def test_mutually_exclusive_options_no_error(set1, set2):
    option1 = OverrideableOption(
        False, "--option1", mutually_exclusive=["option_1, option_2"]
    )
    option2 = OverrideableOption(
        False, "--option2", mutually_exclusive=["option_1, option_2"]
    )
    app = Typer()

    @app.command()
    def _(option_1: bool = option1(), option_2: bool = option2()):
        return "ok"

    command = []
    if set1:
        command.append("--option1")
    if set2:
        command.append("--option2")
    runner = CliRunner()
    result = runner.invoke(app, command)
    print(result.output)
    assert result.exit_code == 0


def test_mutually_exclusive_options_error(snapshot):
    option1 = OverrideableOption(
        False, "--option1", mutually_exclusive=["option_1", "option_2"]
    )
    option2 = OverrideableOption(
        False, "--option2", mutually_exclusive=["option_1", "option_2"]
    )
    app = Typer()

    @app.command()
    def _(option_1: bool = option1(), option_2: bool = option2()):
        return "ok"

    command = ["--option1", "--option2"]
    runner = CliRunner()
    result = runner.invoke(app, command)
    assert result.exit_code == 1
    assert result.output == snapshot


def test_overrideable_option_invalid_callback_signature():
    pass
