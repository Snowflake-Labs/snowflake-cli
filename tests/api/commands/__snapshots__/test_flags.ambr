# serializer version: 1
# name: test_format
  '''
  Usage: default object stage list [OPTIONS] STAGE_NAME
  Try 'default object stage list --help' for help.
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Invalid value for '--format': 'invalid_format' is not one of 'TABLE',        │
  │ 'JSON'.                                                                      │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_mutually_exclusive_options_error
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Options '--option2' and '--option1' are incompatible.                        │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_overrideable_option_is_overrideable
  '''
                                                                                  
   Usage: - [OPTIONS]                                                             
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --option                    INTEGER                  [default: (dynamic)]    │
  │ --install-completion        [bash|zsh|fish|powershe  Install completion for  │
  │                             ll|pwsh]                 the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion           [bash|zsh|fish|powershe  Show completion for the │
  │                             ll|pwsh]                 specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                                               Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
